### PR TITLE
Fix optAssertionProp_Update.

### LINKED ISSUE #24185: Assertion failed 'result == lateArgs->gtOp.gtOp1'
Happens in CI:

https://ci.dot.net/job/dotnet_coreclr/job/master/view/x86/job/jitstress/job/x86_checked_windows_nt_corefx_jitstress1/785/consoleText

Doesn't require JitStress. I repro'ed locally (after setup using `run-corefx-tests.py`):
```
f:\gh\coreclr11\_\fx_x86\artifacts\bin\tests\System.Threading.Tasks.Tests\netcoreapp-Windows_NT-Release-x86>RunTests.cmd --runtime-path f:\gh\coreclr11\_\fx_x86\artifacts\bin\testhost\netcoreapp-Windows_NT-Release-x86 --dotnet-root f:\gh\coreclr11\_\fx_x86\.dotnet
----- start 22:50:37.71 ===============  To repro directly: =====================================================
pushd f:\gh\coreclr11\_\fx_x86\artifacts\bin\tests\System.Threading.Tasks.Tests\netcoreapp-Windows_NT-Release-x86\
set COMPlus_TieredCompilation=0
"f:\gh\coreclr11\_\fx_x86\artifacts\bin\testhost\netcoreapp-Windows_NT-Release-x86\dotnet.exe" xunit.console.dll System.Threading.Tasks.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=nonnetcoreapptests -notrait category=nonwindowstests -notrait category=IgnoreForCI -notrait category=failing -notrait category=OuterLoop
popd
===========================================================================================================

f:\gh\coreclr11\_\fx_x86\artifacts\bin\tests\System.Threading.Tasks.Tests\netcoreapp-Windows_NT-Release-x86>set COMPlus_TieredCompilation=0

f:\gh\coreclr11\_\fx_x86\artifacts\bin\tests\System.Threading.Tasks.Tests\netcoreapp-Windows_NT-Release-x86>"f:\gh\coreclr11\_\fx_x86\artifacts\bin\testhost\netcoreapp-Windows_NT-Release-x86\dotnet.exe" xunit.console.dll System.Threading.Tasks.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=nonnetcoreapptests -notrait category=nonwindowstests -notrait category=IgnoreForCI -notrait category=failing -notrait category=OuterLoop
  Discovering: System.Threading.Tasks.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Threading.Tasks.Tests (found 539 of 715 test cases)
  Starting:    System.Threading.Tasks.Tests (parallel test collections = on, max threads = 8)
    System.Threading.Tasks.Tests.TaskSchedulerTests.GetTaskSchedulersForDebugger_DebuggerAttached_ReturnsAllSchedulers [SKIP]
      Condition(s) not met: "DebuggerIsAttached"
    System.Threading.Tasks.Tests.TaskSchedulerTests.GetScheduledTasksForDebugger_DebuggerAttached_ReturnsTasksFromCustomSchedulers [SKIP]
      Condition(s) not met: "DebuggerIsAttached"

Assert failure(PID 22808 [0x00005918], Thread: 6384 [0x18f0]): Assertion failed 'result == lateArgs->gtOp.gtOp1' in 'System.Threading.Tasks.Tests.TaskRtTests_Core:RunTaskWaitTest_NegativeTests()' (IL size 667)

    File: f:\gh\coreclr11\src\jit\gentree.cpp Line: 7962
    Image: f:\gh\coreclr11\_\fx_x86\artifacts\bin\testhost\netcoreapp-Windows_NT-Release-x86\dotnet.exe

----- end 22:51:41.72 ----- exit code -1073740286 ----------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by sandreenko**
736291405a: Add few comments and format `gtGetThisArg`.

332bd2ec8a: Change `gtFindLink` to return the parent in addition to the link.

6418fcf88c: Fix parent's links in `fgArgInfo` in optAssertionProp_Update. <- **this change fixes the issue.**

e5decb744c: Add a repro test.

0bf52e4304: Use `ReplaceOperand`.

Fixes #24185

### CODE REVIEW

**Comment by sandreenko** [comment]
x64 shows expected diffs on the repro test,
from:
```
IN0030: 000111 mov      rsi, rdx
IN0031: 000114 mov      rdx, rsi
IN0032: 000117 mov      rcx, 0x7FFE73491B18
IN0033: 000121 call     CORINFO_HELP_ISINSTANCEOFCLASS
IN0034: 000126 test     rax, rax
IN0035: 000129 je       SHORT G_M31095_IG09
IN0036: 00012B mov      rcx, rsi
IN0037: 00012E cmp      dword ptr [rcx], ecx <- this cmp could be deleted.
IN0038: 000130 call     System.AggregateException:get_InnerExceptions():ref:this
IN0039: 000135 mov      rcx, rax
IN003a: 000138 xor      edx, edx
IN003b: 00013A cmp      dword ptr [rcx], ecx
IN003c: 00013C call     System.Collections.ObjectModel.ReadOnlyCollection`1[__Canon][System.__Canon]:get_Item(int):ref:this
IN003d: 000141 mov      rdx, rax
IN003e: 000144 mov      rcx, 0x7FFE73331738
IN003f: 00014E call     CORINFO_HELP_ISINSTANCEOFCLASS
IN0040: 000153 test     rax, rax
IN0041: 000156 jne      SHORT G_M31095_IG11
IN0042: 000158 mov      rcx, rsi
IN0043: 00015B cmp      dword ptr [rcx], ecx <- this cmp could be deleted.
```

to:
```
 IN0030: 000111 mov      rsi, rdx
IN0031: 000114 mov      rdx, rsi
IN0032: 000117 mov      rcx, 0x7FFE70901B18
IN0033: 000121 call     CORINFO_HELP_ISINSTANCEOFCLASS
IN0034: 000126 test     rax, rax
IN0035: 000129 je       SHORT G_M31095_IG09
IN0036: 00012B mov      rcx, rsi
                                            
IN0037: 00012E call     System.AggregateException:get_InnerExceptions():ref:this
IN0038: 000133 mov      rcx, rax
IN0039: 000136 xor      edx, edx
IN003a: 000138 cmp      dword ptr [rcx], ecx
IN003b: 00013A call     System.Collections.ObjectModel.ReadOnlyCollection`1[__Canon][System.__Canon]:get_Item(int):ref:this
IN003c: 00013F mov      rdx, rax
IN003d: 000142 mov      rcx, 0x7FFE707A1738
IN003e: 00014C call     CORINFO_HELP_ISINSTANCEOFCLASS
IN003f: 000151 test     rax, rax
IN0040: 000154 jne      SHORT G_M31095_IG11
IN0041: 000156 mov      rcx, rsi
                      
```
where 3 cmp are deleted in the diff because ` CORINFO_HELP_ISINSTANCEOFCLASS` has already checked them.

**Comment by sandreenko** [comment]
test Windows_NT arm64 Cross Checked Innerloop Build and Test

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x64 Checked pmi_asm_diffs
@dotnet-bot test Windows_NT x86 Checked pmi_asm_diffs

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x64 Checked pmi_asm_diffs
@dotnet-bot test Windows_NT x86 Checked pmi_asm_diffs

**Comment by sandreenko** [comment]
No asm diffs on System.Private.CoreLib (x64/x86, checked, pmi/default).

**Comment by sandreenko** [comment]
PTAL @CarolEidt @dotnet/jit-contrib

File: src/jit/assertionprop.cpp [hunk 1]
```diff
@@ -3912,6 +3913,26 @@ GenTree* Compiler::optAssertionProp_Update(GenTree* newTree, GenTree* tree, GenT
             // optAssertionPropMain(). It will reset the gtPrev and gtNext links for all nodes.
             newTree->gtNext = tree->gtNext;
 
+            if ((linkData.parent != nullptr) && linkData.parent->IsCall())
+            {
+                GenTreeCall* parentCall = linkData.parent->AsCall();
```
Comment (CarolEidt) [inline src/jit/assertionprop.cpp]: I would prefer to see this leverage the code in `GenTreeCall::ReplaceCallOperand`. Although it is primarily used in `LIR`, I don't think it depends on it, and it is a more thorough handling of calls. I'm not crazy about the use of `TryGetUse`: https://github.com/dotnet/coreclr/blob/master/src/jit/gentree.cpp#L1041
```
    assert(TryGetUse(*useEdge, &useEdge));
```
which overwrites its argument without verifying that it gets the same result (and `TryGetUse` is primarily an `LIR` thing, but basically just searches for a parent edge). But I don't think we should add this duplicate functionality.

**Comment by sandreenko** [comment]
> Looks great - thanks!

Thanks @CarolEidt for your help and review. 

I had to push another change because `gtFindLink` starts `fgWalkTreePre` with `&stmt->gtStmtExpr` and does not set parent for  `gtStmtExpr`. So I did as rationalize does:
https://github.com/dotnet/coreclr/blob/c3e404425f9f03088f0e7106914190634baef619/src/jit/rationalize.cpp#L165-L175

Note: `GT_STMT ` was intentionally made as non-valid parent: 
https://github.com/dotnet/coreclr/blob/72e450b5d4eb6cc8fd9e99e0202b538b795350e8/src/jit/gentree.cpp#L4538-L4541

https://github.com/dotnet/coreclr/blob/72e450b5d4eb6cc8fd9e99e0202b538b795350e8/src/jit/gentree.cpp#L4683-L4686

**Comment by CarolEidt** [comment]
> Note: GT_STMT was intentionally made as non-valid parent:

That's probably the LIR impact - i.e. there's no `GT_STMT`s in LIR, and in LIR you can have regular tree nodes with no parent, unlike in HIR, where you always have a stmt, so you could theoretically handle this without worrying about parent being null in the HIR case. One might consider handling stmt nodes, but you'd also have to return the stmt node from `gtFindLink` which would force it to start at that node, so this is probably better as-is.

**Comment by sandreenko** [comment]
Build tests failures are covered by https://github.com/dotnet/roslyn/issues/35290

**Pull request merged**